upbound/up
# Add `up enterprise` commands and refactor helm installer

### Description of your changes



From a user-facing perspective, this PR adds support for `up enterprise ...` commands, but in order to enable that, it refactors the helm installer to be more general and adds a drop-in helm pull client that can pull charts from an OCI registry. The more flexible helm installer means that we can continue to use it for other charts we may want to install in the future, and we can add features for our existing installation targets (`uxp` and `enterprise`) without needing to special case them. There are likely some additional ergonomic improvements that can be made to the installer, and I imagine there will be some additional tweaks over time.

I am finishing up writing some unit tests for `registryPuller`, but this PR is otherwise good to go. See testing steps below.

Fixes #112 

I have:


### How has this code been tested



I have extensively tested each of the following scenarios:
- Installing Crossplane then upgrading to compatible UXP version
```
$ helm install crossplane --namespace crossplane-system crossplane-stable/crossplane
<observe healthy install>
$ up uxp upgrade v1.4.0-up.1 -n crossplane-system
<observe healthy install>
```
- Install latest stable UXP version, then upgrade to newer unstable version, then uninstall
```
$ up uxp install
<observe healthy install>
$ up uxp upgrade --unstable 1.5.0-rc.0.up.1.4.g2823565
<observe healthy upgrade>
$ up uxp uninstall
<observe all components removed>
```
- Install enterprise version, overriding with hidden repo arg, then upgrade to newer version, then uninstall
```
$ up enterprise install --repo registry.upbound.io/enterprise-dev v0.0.0-110.gcc7d34d --set withCertManager=false
License ID: <REDACTED>
Token: 
<observe healthy install -- i.e. all components are there but images fail to be pulled because I did not create image pull secret>
$ up enterprise upgrade --repo registry.upbound.io/enterprise-dev v0.0.0-112.g63a2b32 --set withCertManager=false
License ID: <REDACTED>
Token: 
<observe healthy upgrade -- i.e. all components are there but images fail to be pulled because I did not create image pull secret>
$ up enterprise uninstall
<observe all components removed>
```
